File: RealTime/ARC036/arc036c.py
# 改良に改良を重ねたけど限界
# JavaでのAC：http://arc036.contest.atcoder.jp/submissions/380640
# その他の提出
# https://arc036.contest.atcoder.jp/submissions/380642
# https://arc036.contest.atcoder.jp/submissions/380635
# https://arc036.contest.atcoder.jp/submissions/380629
# https://arc036.contest.atcoder.jp/submissions/380625
# https://arc036.contest.atcoder.jp/submissions/380615
# https://arc036.contest.atcoder.jp/submissions/380593
# https://arc036.contest.atcoder.jp/submissions/380567

import copy

N, K = map(int, input().split())
slist = list(input())
MODNUM = 1000000007

dp = [[0 for _ in range(K + 2)] for _ in range(K + 2)]
ndp = copy.deepcopy(dp)
dp[0][0] = 1

for t in range(0, N):
    casii = ord(slist[t])
    for i in range(0, K + 1):
        for j in range(0, K + 1):
            if dp[i][j] == 0:
                continue

            if casii == 48 or casii == 63:
                nj = j - 1 if j > 0 else 0
                ndp[i + 1][nj] += dp[i][j]
            if casii == 49 or casii == 63:
                ni = i - 1 if i > 0 else 0
                ndp[ni][j + 1] += dp[i][j]

    for i in range(0, K + 1):
        for j in range(0, K + 1):
            dp[i][j] = ndp[i][j]
            ndp[i][j] = 0

res = 0
for i in range(0, K + 1):
    res += sum(dp[i]) - dp[i][K + 1]

print(res % MODNUM)

File: RealTime/ARC036/arc036c.java
import java.util.Scanner;
 
class Main
{
	public static void main (String[] args)
	{
		Scanner scan = new Scanner(System.in);
		int N = scan.nextInt();
		int K = scan.nextInt();
		String s = scan.next();
		
		int MODNUM = 1000000007;
		int[][][] dp = new int[N + 1][K + 2][K + 2];
		dp[0][0][0] = 1;
 
		for(int t = 0; t < N; t++) {
			char c = s.charAt(t);
			for(int i = 0; i < K + 1; i++) {
				for(int j = 0; j < K + 1; j++) {
					if(dp[t][i][j] == 0)
                		continue;
		
        		    if(c == '0' || c == '?') {
                		int nj = j > 0 ? j - 1 : 0;
                		dp[t + 1][i + 1][nj] += dp[t][i][j];
                		if(dp[t + 1][i + 1][nj] >= MODNUM) dp[t + 1][i + 1][nj] -= MODNUM;
        		    }
            		if(c == '1' || c == '?') {
                		int ni = i > 0 ? i - 1 : 0;
                		dp[t + 1][ni][j + 1] += dp[t][i][j];
                		if(dp[t + 1][ni][j + 1] >= MODNUM) dp[t + 1][ni][j + 1] -= MODNUM;
            		}
				}
			}
		}
		
		int res = 0;
		for(int i = 0; i < K + 1; i++) {
			for(int j = 0; j < K + 1; j++) {
				res += dp[N][i][j];
				if(res >= MODNUM) res -= MODNUM;
			}
		}
		
		res %= MODNUM;
		System.out.println(res);
	}
}